ceph: 12.2.7->13.2.4

krav opened the pull request:
###### Motivation for this change

This upgrades ceph to the latest version. Closes #42830.

There are caveats:
  * The dashboard is not included. This seems to be in line with other distributions, upstream master has a less byzantine build system for it coming.
  * The package builds its own boost, I couldn't get it to locate ours.
  * I ended up making changes to generic.nix that probably would break ceph 12, and I didn't see a reason to maintain several simultaneous versions, so I merged it with default.nix (in it's own commit).

Additionally I added a `client` output containing just enough to manage `rbd` mounts (modelled on debian's `ceph-common`) and fixed some permission and path stuff in the module (see commit).

Ping @nh2, @wkennington, @adev, @ak3n.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

krav committed:
ceph: 12.2.7->13.2.2

krav committed:
ceph: maintain only one version

krav committed:
nixos/ceph: extraConfig, fix test, wait for ceph-mgr to become active, ..

* run ceph with disk group permission
* fix crashing radosgw by correct path to mime-types
* add extraConfig option for the global section
* define users using users.extraUsers instead of users.users
* remove assertion that mgr be enabled on /every/ node, only one is
needed per cluster

* more descriptive function names
* clear up how ceph.conf is generated

nixos/ceph: Remove extranous extras. Add examples, types.

Add example for extraConfig, path type for mime option,
line breaks and explanations.

nixos/ceph: add wantedBy=ceph.target to daemon targets

nixos/ceph: Fix ceph testcase

As services.ceph.enable also enables the ceph.target, it makes the testcase fail
due to race-conditions before bootstraping. Thus we force the wantedBy of the
ceph target to an empty list to stop it from being started during bootup. To
ensure that ceph.target still work as intended, all daemons are stopped
at the end of the test and then started again with ceph.target, with
verification steps to ensure the cluster comes up as it should too.

nixos/ceph: Add wait for ceph-mgr to come up and active

The ceph-mgr takes some time before it is up and running correctly. When it is done
its marked as active in the status output, this is needed for proper function of the
cluster. To prevent the race-condition of the ceph-mgr daemon coming up and the
time it takes to bootstrap the OSDs in the tests before starting them in the
future, a wait for the ceph-mgr daemon to be ready was added.
This commit also removes an unneeded folder path creation.

krav committed:
ceph: 13.2.2 -> 13.2.4, fixes

ceph: erroneous comma, fix target dependencies

ceph: fix lttng-ust alias

ceph: Use system boost

ceph: Add documentation on python27 usage

ceph: substitute substituteInPlace for patch, init ceph-client

ceph: include ceph-volume python tool in output

ceph: fix rados python module path

ceph: Fix callPackage to callPackages

The function callPackage masks the override attribute and makes ceph
non-overridable by the user. The function callPackages fixes this
by adding the override attribute on each set.

ceph: 13.2.2 -> 13.2.4

ceph: remove argparse

ceph: remove wkennington as maintainer

srhb commented:
cc @lejonet if you're still Ceph-knowledgeable. :)

krav commented on nixos/modules/services/network-filesystems/ceph.nix:
```diff
@@ -313,54 +328,54 @@ in
       uid = config.ids.uids.ceph;
       description = "Ceph daemon user";
+      group = "ceph";
+      extraGroups = [ "disk" ];
```
Most of these `extra`s are caused by debugging while learning NixOS, but this one is needed. As far as I can tell there is no `users.groups`, and auxiliary groups are specified in `users.extraGroups`. 

The `osd` daemon needs `disk` permissions as it's the component that accesses disks, which it why it also has `PrivateDevices = "no"` set below. The others will happily run as `ceph:ceph`. I see now that every daemon runs as `ceph:disk`, I'll fix that. (Specifying the `--group <group>` option will drop all other group privileges, so as long as the systemd service starts as `root` I can't see how to make ceph run with both groups).

Thank you for your extensive feedback, I'll make a commit addressing it soon.

krav commented on pkgs/tools/filesystems/ceph/0000-fix-SPDK-build-env.patch:
```diff
@@ -0,0 +1,11 @@
+     # its include directory.
+     # unset $LDFLAGS, otherwise SPDK will fail to mock some functions.
+-    BUILD_COMMAND env -i PATH=$ENV{PATH} CC=${CMAKE_C_COMPILER} $(MAKE) EXTRA_CFLAGS="-fPIC"
++    BUILD_COMMAND env PATH=$ENV{PATH} CC=${CMAKE_C_COMPILER} $(MAKE) EXTRA_CFLAGS="-fPIC" C_OPT="-mssse3"
```
They clean up the environment here a bit too well, removing some Nix specific thing needed to build. I forget what, of course, I'll see if I can't reconstruct it and leave a comment. Our environment doesn't include any interfering LDFLAGS, so starting afresh is unnecessary. The `-mssse3` flag is courtesy of the nixpkgs build of spdk, which rightfully notes it won't build without it.

krav commented on pkgs/tools/filesystems/ceph/default.nix:
```diff
@@ -1,12 +1,173 @@
+  meta = {
+    homepage = https://ceph.com/;
+    description = "Distributed storage system";
+    license = licenses.lgpl21;
```
Is there a nixpkgs way of specifying compound licensing? It's mostly LGPL and more permissive licenses, but the README notes "There are a handful of headers included here that are licensed under the GPL". Should the least permissive win out?

srhb replied:
`license` can be a list. For the purpose of (not) building on Hydra, if any is unfree, we handle that correctly. There is no special handling of multiple free ones, it *should* just work.

krav replied:
I've added a bunch. Some of the documentation is CC-BY-SA, but I don't think any of it ends up in our build, so I've omitted that.

lejonet commented:
I've gone through the PR and see no shenaningans you haven't addressed. I fully agree with the name changes on the variables in the service module and thank you for removing the mgr requirement that I have forgotten to fix with a PR of my own (@srhb and I have talked about that one before, it was a naive assumption from my part when making the module, because I was only testing single node configuration).

I'd ofc also like an answer of the extraUsers vs users, because from my understanding, they should be the same and equivalent to each other but I never found any docs telling which should be preferred, which is why I chose the ones without extra.

lejonet commented on nixos/modules/services/network-filesystems/ceph.nix:
```diff
@@ -157,6 +159,22 @@ in
+    };
+
+    extraConfig = mkOption {
+      type = with types; attrsOf str;
```
This is left over from me, there is an example for the OSD daemon, that probably should be done for each daemon type to not require people to actually know that (they might setup the OSDs differently or not with this module for example)

krav replied:
I've added an example to the global extraConfig, additionally, as lejonet mentioned, there's one for the OSD section. I see that it might be unintuitive for users, but in return you get a bit of syntax checking and programmatic ease, which might be useful for some people.

lejonet reviewed (commented):
Biggest thing to note regarding maintaining only one version is that every other major ceph release is a LTS release. This means that all security fixes get backported to that release until the next LTS release. So there is an argument for having 1 LTS package and one "current" package for ceph, but as they are going through some changes with the build-env and such, to make it much easier to package and maintain, just having one release atm makes sense.

krav commented on nixos/modules/services/network-filesystems/ceph.nix:
```diff
@@ -157,6 +159,22 @@ in
       };
+
+      rgwMimeTypesFile = mkOption {
+        type = with types; nullOr str;
```
Fixed.

krav commented on nixos/modules/services/network-filesystems/ceph.nix:
```diff
@@ -313,54 +328,54 @@ in
         generators.toINI {} totalConfig;
 
-    users.users = singleton {
+    users.extraUsers = singleton {
```
No reason :) Fixed.

krav commented on pkgs/tools/filesystems/ceph/0000-dont-hardcode-bin-paths.patch:
```diff
@@ -0,0 +1,22 @@
+ 	char command[128];
+ 
+ 	snprintf(command, sizeof(command),
+-		 "/sbin/modinfo -F parm %s | /bin/grep -q ^%s:",
```
Absolutely, fixed.

krav commented on pkgs/tools/filesystems/ceph/default.nix:
```diff
@@ -1,12 +1,173 @@
+    homepage = https://ceph.com/;
+    description = "Distributed storage system";
+    license = licenses.lgpl21;
+    maintainers = with maintainers; [ adev ak wkennington ];
```
Fine :) Added.

krav commented on pkgs/tools/filesystems/ceph/default.nix:
```diff
@@ -1,12 +1,173 @@
+
+  enableParallelBuilding = true;
+
+  outputs = [ "out" "lib" "dev" "doc" "client" "man" ];
```
I've made it a separate derivation, I think that turned out a lot better. Thanks for the idea!

krav commented:
I'm putting a `wip` on this. Just discovered that `ceph-volume` has disappeared from the output, and I bet some people find that tool useful.

srhb commented:
Outstanding issues:

- Check that things that depend on ceph still work after the changes to all-packages
- Rebase on master/unstable (and remove the extraneous argparse dep)
- ?

srhb commented:
@krav Do you want to address the points above, or would you rather one of lejonet or I do it? :) I don't want to force push to this branch without your go-ahead.

krav commented:
@srhb I'm easily confused by git, but I've rebased and squashed now. 

Regarding your first point, what things other things depend on Ceph and need checking?

srhb commented:
@krav Hm, looks like nothing by default. There's a bunch of things that has it as an optional dependency, but we don't have any relevant tests for that.

@lejonet has done extensive testing and thinks we're good.

Status: I know of just one problem left, and that's the dashboard module failing to get the version from cherrypy. I suspect it does not in fact get launched with the correct python env, but I need to actually investigate this.
Surprisingly, due to the way the release tarball is made, we're fine on the frontend side, so it really is just the dashboard backend being broken.

srhb commented:
This (sledgehammer) appears to work, but I'd really prefer to figure out why cherrypy reports a wrong version, but only in the (weird?) ceph env.

https://github.com/srhb/nixpkgs/commit/ef3a3807f75719f26563128e0bce5dc1b0afe64f

lejonet commented:
> This (sledgehammer) appears to work, but I'd really prefer to figure out why cherrypy reports a wrong version, but only in the (weird?) ceph env.
> 
> [srhb@ef3a380](https://github.com/srhb/nixpkgs/commit/ef3a3807f75719f26563128e0bce5dc1b0afe64f)

I will see if I can get some digging done on it tonight, I agree with you that even if that sledgehammer works, it feels appropriate to find the root cause of it. So that we can fix the actual problem instead of hiding the symptoms.

azazel75 commented:
guys, @lejonet: any news on this?

lejonet commented:
I sadly was hit with a lot of real life so I haven't been able to do more digging than we did at this time, @srhb might have done more but iirc theu also were hit with real life

johanot commented:
@srhb @lejonet I finally managed to get some time at work allocated for ceph. Where do you suggest I start? I guess we might just jump directly to Ceph 14 now, but question is whether we should try and get this PR merged first thing?

lordcirth commented:
While I personally would prefer to have Nautilus, Mimic is an LTS release, so there's an argument to be made for focusing on getting that working instead.

johanot commented:
@lordcirth Oh.. I thought Nautilus was LTS. Appently not. In that case, I agree with you.

lordcirth commented:
> @lordcirth Oh.. I thought Nautilus was LTS. Appently not. In that case, I agree with you.
Yes, it's noted here in the .0 changelogs:

http://docs.ceph.com/docs/master/releases/nautilus/#v14-2-0-nautilus
http://docs.ceph.com/docs/master/releases/mimic/#v13-2-0-mimic

azazel75 commented:
I've spent some time in testing this PR out and it builds issue, but then I haven't tried a real setup and I stopped looking into this ....I was trying to understand if that problem with the dashboard is still there but I'm kind of a newbie on ceph and there is a ton of stuff in here ;-).. anyway, if I can be of any help, let me know

lordcirth commented:
v13.2.6 just landed:

```
Notable Changes
---------------
* Ceph v13.2.6 now packages python bindings for python3.6 instead of
  python3.4, because EPEL7 recently switched from python3.4 to
  python3.6 as the native python3. See the announcement[1] _`
  for more details on the background of this change.

```

lejonet commented:
@johanot as @lordcirth noted, mimic is a LTS and thus there is value to ensure that this PR gets merged first. Mainly because it comes with some needed "upgrades" to the service module, in my own opinion at least, and secondly because its more or less done.

"The only thing" that needs to be fixed iirc is the dashboard plugin. The problem there is that they are using an embedded python interprenter, which they instrument through the CPython API. This means that our munging with env vars and so, to get stuff to find the relevant python env for it, doesn't have any effect.
This in turn results in the env being lost between the loading of a module and executing it, which leads to the dashboard plugins usage of cherrypy failing because it cannot find the .egg file to do a version check that enables some monkeypatching on very old version of cherrypy.

With that said, in 14.2 they have refactored the dashboard plugin to be its own thing, and thus isn't "a part of" standard ceph anymore so I would suggest that minor work should be done with this PR to get it to 13.2.6. Potentially using the sledgehammer approach that @srhb found in https://github.com/NixOS/nixpkgs/pull/49866#issuecomment-469406445

Work should then be done to get us up to using python3 and potentially 14.2.x (might be possible in the same PR, I haven't scrutinized the changes between 13 and 14).

johanot commented:
Ok.. I've managed to bump to 13.2.6 here: https://github.com/NixOS/nixpkgs/compare/master...johanot:ceph-13 .. But unfortunately it required applying the sledgehammer a couple of times again, removing two more cherrypy version checks. :( (https://github.com/johanot/nixpkgs/commit/153479459f0d7cd433d927804e8f53ed38c8ac6d)

Without hesitation I've deployed this branch to our production env. Will try to test it sufficiently during the coming week.

johanot commented:
The news from my side is that we are now on ceph 14 in production. The ceph 14 upgrade (which I have on another branch somewhere) is largely based on the changes in this PR, and while we ran ceph 13 (https://github.com/johanot/nixpkgs/commit/153479459f0d7cd433d927804e8f53ed38c8ac6d) in prod, we saw no issues. So basically, I'm ok with my changes getting merged, sledgehammering included, but... I neither have permissions to push nor merge.

devhell commented:
@johanot Maybe, if you have time, it would be good to open a separate PR for Ceph 14 and reference this PR?

grahamc commented on nixos/tests/ceph.nix:
```diff
@@ -1,4 +1,4 @@
+import ./make-test.nix ({pkgs, lib, ...}: rec {
   name = "All-in-one-basic-ceph-cluster";
   meta = with pkgs.stdenv.lib.maintainers; {
     maintainers = [ lejonet ];
```
Would you like to be a maintainer?

lejonet replied:
I guess I should add myself to maintainers. So yes, I want to be a maintainer.

grahamc commented on nixos/modules/services/network-filesystems/ceph.nix:
```diff
@@ -157,6 +163,27 @@ in
+
+      rgwMimeTypesFile = mkOption {
+        type = with types; nullOr path;
+        default = "${pkgs.mime-types}/etc/mime.types";
```
This should use defaultText: https://github.com/NixOS/nixpkgs/blob/2e869f71a7aebf0e96114cf7317b73583e893eac/nixos/modules/config/krb5/default.nix#L86-L87

grahamc commented on nixos/tests/ceph.nix:
```diff
@@ -51,6 +51,9 @@ import ./make-test.nix ({pkgs, ...}: rec {
       };
+
+      # So that we don't have to battle systemd when bootstraping
+      systemd.targets.ceph.wantedBy = lib.mkForce [];
```
Does this mean bootstrapping a new cluster is tricky?

lejonet replied:
Both yes and no. The main reason I force the target off in the test is to not get a bunch of errors related to daemons not being able to find their files, because they haven't been bootstrapped yet, to clutter the test output. The errors are harmless but made it significantly harder to find actual errors in the output.

johanot commented:
@grahamc @srhb Could one of you please rebase this branch with master and perhaps my ceph-13 branch as well? Then we can do another round of integration-testing and a collective review of all the changes.

srhb commented:
I can do that tonight if you don't just want to open a separate PR with those changes. :)

johanot commented:
@srhb as we talked about in another channel, I'll try to assemble a new branch and PR it separately. I can say already that ceph 13.2.6 doesn't build after rebasing with master :P I'll be back at some point once I get that sorted.

johanot commented:
TLDR; there's a collision on python2 backports on master, preventing ceph from building. I decided that instead of wasting time fiddling around with the python2-env, perhaps we should try to have a go with ceph 14, which uses python3. Please see #65470 and please comment and review over there.

srhb commented:
Included in #65470 -- thanks again

nh2 commented:
> "The only thing" that needs to be fixed iirc is the dashboard plugin. The problem there is that they are using an embedded python interprenter, which they instrument through the CPython API. This means that our munging with env vars and so, to get stuff to find the relevant python env for it, doesn't have any effect.
> This in turn results in the env being lost between the loading of a module and executing it

@lejonet Can you explain this a bit further?

I'm trying to get the dashboard to work with Ceph 15.2.3, and get:

```
10 mgr[py] Computed sys.path '/nix/store/ahn8d1bnkrvj7gbxc90cwqq037g740rs-ceph-15.2.3/lib/python3.7/site-packages:/nix/store/4naw242f2izj1rs55hxcc66zifqcqbn0-python3-3.7.6-env/lib/python3.7/site-packages:/nix/store/55b9ip7xkpimaccw9pa0vacy5q94f5xa-python3-3.7.6/lib/python37.zip:/nix/store/55b9ip7xkpimaccw9pa0vacy5q94f5xa-python3-3.7.6/lib/python3.7:/nix/store/55b9ip7xkpimaccw9pa0vacy5q94f5xa-python3-3.7.6/lib/python3.7/lib-dynload:/nix/store/icpiyydgbnp62pvdh96ff0g4by2zy145-ceph-unwrapped-15.2.3/share/ceph/mgr:/nix/store/55b9ip7xkpimaccw9pa0vacy5q94f5xa-python3-3.7.6/lib/python3.7/site-packages'
-1 mgr[py] Module not found: 'cephadm'
-1 mgr[py] Traceback (most recent call last):
  File "/nix/store/icpiyydgbnp62pvdh96ff0g4by2zy145-ceph-unwrapped-15.2.3/share/ceph/mgr/cephadm/__init__.py", line 6, in <module>
    from .module import CephadmOrchestrator
  File "/nix/store/icpiyydgbnp62pvdh96ff0g4by2zy145-ceph-unwrapped-15.2.3/share/ceph/mgr/cephadm/module.py", line 31, in <module>
    from ceph.deployment import inventory, translate
ModuleNotFoundError: No module named 'ceph'

-1 mgr[py] Class not found in module 'cephadm'
```

which is [this](https://github.com/ceph/ceph/blob/9fd2f65f91d9246fae2c841a6222d34d121680ee/src/mgr/PyModule.cc#L673-L679).

I do not understand why it cannot find the `ceph` module, because it uses [`PySys_SetPath(get_site_packages())`](https://github.com/ceph/ceph/blob/9fd2f65f91d9246fae2c841a6222d34d121680ee/src/mgr/PyModule.cc#L342-L345) which correctly includes `/nix/store/...-ceph-15.2.3/lib/python3.7/site-packages/` (which contains `ceph-1.0.0-py3.7.egg/ceph/__init__.py`), and if I set `PYTHONPATH` to this for my `python3` and `import ceph`, that works just fine.

What does the CPython API do different that results in this breakage?

nh2 commented:
I can see in `strace` that the maximum it does with `ceph-1.0.0-py3.7.egg` is to read the contents of this dir:

```c
getdents64(22</nix/store/...-ceph-15.2.3/lib/python3.7/site-packages/ceph-1.0.0-py3.7.egg>, [{d_ino=1317894, d_off=783223069217015699, d_reclen=32, d_type=DT_DIR, d_name="EGG-INFO"}, {d_ino=1194113, d_off=3081453319264304844, d_reclen=24, d_type=DT_DIR, d_name=".."}, {d_ino=1317901, d_off=5351582700327092632, d_reclen=24, d_type=DT_DIR, d_name="ceph"}, {d_ino=1317893, d_off=9223372036854775807, d_reclen=24, d_type=DT_DIR, d_name="."}], 32768) = 104
getdents64(22</nix/store/...-ceph-15.2.3/lib/python3.7/site-packages/ceph-1.0.0-py3.7.egg>, [], 32768) = 0
close(22</nix/store/...-ceph-15.2.3/lib/python3.7/site-packages/ceph-1.0.0-py3.7.egg>) = 0
```

But then it never does more than that, it never goes into `ceph-1.0.0-py3.7.egg/ceph/`.

nh2 commented:
Looks like I can work around it by putting `$(toPythonPath $out)/ceph-1.0.0-py3.7.egg` straight onto the `PYTHONPATH`, I'm not sure why that's necessary though, and it's ugly.

nh2 commented:
Also I found that some nixpkgs python packages do

```sh
rm -f "$out/lib/${python.libPrefix}"/site-packages/site.py*
```

but i don't know why and none of them have comments on it. @FRidh @jonringer Do you know?

Maybe it's related?

nh2 commented:
When I run (tip from `energizer` on IRC):

```
# PYTHONPATH=/nix/store/1fy23hzpll8kx160z1mb502bq7rrirzc-ceph-15.2.3/lib/python3.7/site-packages/ python3 -m site
sys.path = [
    '/root',
    '/nix/store/1fy23hzpll8kx160z1mb502bq7rrirzc-ceph-15.2.3/lib/python3.7/site-packages',
    '/nix/store/1fy23hzpll8kx160z1mb502bq7rrirzc-ceph-15.2.3/lib/python3.7/site-packages/ceph_volume-1.0.0-py3.7.egg',
    '/nix/store/1fy23hzpll8kx160z1mb502bq7rrirzc-ceph-15.2.3/lib/python3.7/site-packages/ceph-1.0.0-py3.7.egg',
    '/nix/store/55b9ip7xkpimaccw9pa0vacy5q94f5xa-python3-3.7.6/lib/python37.zip',
    '/nix/store/55b9ip7xkpimaccw9pa0vacy5q94f5xa-python3-3.7.6/lib/python3.7',
    '/nix/store/55b9ip7xkpimaccw9pa0vacy5q94f5xa-python3-3.7.6/lib/python3.7/lib-dynload',
    '/nix/store/55b9ip7xkpimaccw9pa0vacy5q94f5xa-python3-3.7.6/lib/python3.7/site-packages',
]
USER_BASE: '/root/.local' (doesn't exist)
USER_SITE: '/root/.local/lib/python3.7/site-packages' (doesn't exist)
ENABLE_USER_SITE: True
```

That looks like what it needs (putting `ceph-1.0.0-py3.7.egg` onto the path), but I don't understand why that's not in effect, and why it would be bad.

I can also see:

```
# cat /nix/store/1fy23hzpll8kx160z1mb502bq7rrirzc-ceph-15.2.3/lib/python3.7/site-packages/easy-install.pth 
./ceph_volume-1.0.0-py3.7.egg
./ceph-1.0.0-py3.7.egg
```

But I'm not sure what that does.

lejonet commented:
@nh2 It was a while since I did the deep dive into the whole mess that is cephs handling of python modules with Cpython, but if I remember correctly, due to how they initialized the actual interprenter, the PYTHONPATH env variable did not propagate to the actual interprenter that it then starts all modules with. This is a "working as intended" of the Cpython API, iirc it even said in the documentation around the functions they used that if you wanted to preserve env vars, like PYTHONPATH, you had to explicitly set them ahead of invoking the Cpython interprenter.
